### PR TITLE
DOC: Added docs to clarify that MultiIndex.set_levels() interprets passed values as new components of the .levels attribute (#28294)

### LINKED ISSUE #28294: Clarify that MultiIndex.set_levels() interprets passed values as new components of the .levels attribute
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd
df = pd.DataFrame(np.random.rand(3, 3), columns=pd.MultiIndex.from_tuples([(1, 2), (1, 4), (5, 6)]))
df.columns = df.columns.set_levels([1, 1, 3], level=0)
```
Which raises a ```ValueError```:

```python
ValueError: Level values must be unique: [1, 1, 3] on level 0
```

#### Problem description

Despite a dataframe with non-unique MultiIndex can be created, they cannot be set using ```set_levels()```. Is this behaviour expected? I believe non-unique level values were not allowed for a period of time (#18882), but then they were allowed again (#21423), so I am not sure which is the current convention.

#### Expected Output

```python
          1                   3
          2         4         6
0  0.317669  0.329142  0.056725
1  0.969472  0.340309  0.135204
2  0.242408  0.934748  0.683186
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.1
numpy            : 1.16.4
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.2
setuptools       : 41.0.1
Cython           : 0.29.13
pytest           : 5.1.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.1.8
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.8.0
pandas_datareader: None
bs4              : None
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : 2.6.2
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.7
tables           : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
xlsxwriter       : 1.1.8

</details>

### PR DESCRIPTION
**Pull request by hweecat**
Added documentation on MultiIndex.set_levels to clarify that MultiIndex.set_levels() interprets passed values as new components of the .levels attribute

- [x] closes #28294 - or at least attempts to
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Here's my first PR to pandas - looking forward to feedback on how I can further improve this.

### CODE REVIEW

File: doc/source/user_guide/advanced.rst [hunk 1]
```diff
@@ -177,6 +177,37 @@ completely analogous way to selecting a column in a regular DataFrame:
 See :ref:`Cross-section with hierarchical index <advanced.xs>` for how to select
 on a deeper level.
 
+.. _advanced.set_levels:
```
Comment (jreback) [inline doc/source/user_guide/advanced.rst]: I would rather just edit the actual set_levels doc-string with examples. This is too much for the main docs (and not generally what someone would want).

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -740,32 +748,51 @@ def set_levels(self, levels, level=None, inplace=False, verify_integrity=True):
         Examples
         --------
         >>> idx = pd.MultiIndex.from_tuples([(1, 'one'), (1, 'two'),
-                                            (2, 'one'), (2, 'two')],
+                                            (2, 'one'), (2, 'two'),
+                                            (3, 'one'), (3, 'two')],
                                             names=['foo', 'bar'])
-        >>> idx.set_levels([['a', 'b'], [1, 2]])
+        >>> idx.set_levels([['a', 'b', 'c'], [1, 2]])
```
Comment (datapythonista) [inline pandas/core/indexes/multi.py]: Why do you think an extra value makes this example clearer? I think a more concise example is easier to read.

File: pandas/core/indexes/multi.py [hunk 3]
```diff
@@ -740,32 +748,51 @@ def set_levels(self, levels, level=None, inplace=False, verify_integrity=True):
...
+                    ('a', 2),
+                    ('b', 1),
+                    ('b', 2),
+                    ('c', 1),
+                    ('c', 2)],
                    names=['foo', 'bar'])
-        >>> idx.set_levels([['a', 'b'], [1, 2]], level=[0, 1])
+        >>> idx.set_levels([['a', 'b', 'c'], [1, 2]], level=[0, 1]).levels
+        FrozenList([['a', 'b', 'c'], [1, 2]])
+        >>> idx.set_levels([['a', 'b', 'c'], [1, 2, 3, 4]], level=[0, 1])
```
Comment (datapythonista) [inline pandas/core/indexes/multi.py]: Can you separate this example from the previous, and explain what are we showing?

**Review by datapythonista — COMMENTED** [review]
thanks @hweecat for working on this, added some comments.

File: pandas/core/indexes/multi.py [hunk 4]
```diff
@@ -719,8 +719,16 @@ def _set_levels(
 
     def set_levels(self, levels, level=None, inplace=False, verify_integrity=True):
         """
-        Set new levels on MultiIndex. Defaults to returning
-        new index.
+        Set levels on MultiIndex by passing a new value for each
+        index in the level. Defaults to returning new
+        index.
```
[the remark this answers is not in the document]
Comment (hweecat) [reply pandas/core/indexes/multi.py]: Oops, just realised I kinda broke PEP 257 for docstrings. Will make the changes.

File: pandas/core/indexes/multi.py (hunk 2 above)
Comment (hweecat) [reply pandas/core/indexes/multi.py]: My intention of adding an extra value at one of the levels (same for the _MultiIndex_ above) is to further highlight the differences in passing values to the two levels in the _MultiIndex_, for which the number of values in each level are not equal. Using different number of values in each level is to emphasize that _set_levels()_ passes values to indexes on 'by level' basis (by changing the _levels_ attribute), not on 'by row' basis.

Then again, if a more concise example could also explain sufficiently how _set_levels()_ work without ambiguity, I'm okay with reverting to the original _MultiIndex_.

File: pandas/core/indexes/multi.py (hunk 3 above)
Comment (hweecat) [reply pandas/core/indexes/multi.py]: The intention of this example is to illustrate that _set_levels()_ passes values into the _levels_ attribute that is represented by a FrozenList containing list of values for each level in the _MultiIndex_, even when the number of values passed for a level is more than the number of indexes available in the _MultiIndex_ itself.

I'll separate this example from the previous example, and append the explanation to the example.

File: pandas/core/indexes/multi.py [hunk 5]
```diff
@@ -740,32 +748,51 @@ def set_levels(self, levels, level=None, inplace=False, verify_integrity=True):
         Examples
         --------
         >>> idx = pd.MultiIndex.from_tuples([(1, 'one'), (1, 'two'),
-                                            (2, 'one'), (2, 'two')],
+                                            (2, 'one'), (2, 'two'),
+                                            (3, 'one'), (3, 'two')],
                                             names=['foo', 'bar'])
```
[the remark this answers is not in the document]
Comment (hweecat) [reply pandas/core/indexes/multi.py]: Sure, will apply this change.

**Comment by hweecat** [comment]
Travis CI is failing at Python 3.8-dev but passes for Python 3.6 and 3.7 - not too sure how to resolve this.

**Comment by WillAyd** [comment]
@hweecat I'm not sure what happened but the git diff on this seems to have pulled in close to 200 files

Are you still fixing that up? It might be easier to start fresh with a new branch if unsure what happened

**Comment by hweecat** [comment]
I think it could be because I tried to sync my fork with the upstream in order to pass Travis CI for Python 3.8 (Travis CI kept failing for Python 3.8 previously: https://travis-ci.org/pandas-dev/pandas/jobs/600207880).

Think I might need to start fresh with a new branch (from the master that is synced with the latest upstream) and add changes to that branch. Does that mean I might have to create a new pull request?

**Comment by WillAyd** [comment]
Yea it might be easiest to just do that at this point

**Comment by WillAyd** [comment]
Superseded by #29143

**Pull request closed without merging**